ideascube/configuration/management/commands/config.py

```diff
@@ -0,0 +1,153 @@
+import argparse
+import ast
+
+from django.contrib.auth import get_user_model
+from django.core.management.base import BaseCommand, CommandError
+
+from ideascube.configuration import get_config, reset_config, set_config
+from ideascube.configuration.exceptions import (
+    InvalidConfigurationValueError,
+    NoSuchConfigurationKeyError,
+    NoSuchConfigurationNamespaceError,
+    )
+from ideascube.configuration.registry import (
+    get_all_namespaces, get_config_data, get_namespaced_configs,
+    )
+
+
+class Command(BaseCommand):
+    help = 'Manage server configuration'
+
+    def add_arguments(self, parser):
+        subs = parser.add_subparsers(
+            title='Commands', dest='cmd', metavar='',
+            parser_class=argparse.ArgumentParser)
+
+        get = subs.add_parser(
+            'get', help='Get the current value of a configuration option')
+        get.add_argument('namespace', help='The configuration namespace')
+        get.add_argument('key', help='The configuration key')
+        get.set_defaults(func=self.get_config)
+
+        help = subs.add_parser(
+            'help', help='Print help about a configuration option')
+        help.add_argument('namespace', help='The configuration namespace')
+        help.add_argument('key', help='The configuration key')
+        help.set_defaults(func=self.help_for_config)
+
+        list = subs.add_parser(
+            'list', help='List configuration namespaces and keys')
+        list.add_argument(
+            'namespace', nargs='?',
+            help='Only list configuration keys for this namespace')
+        list.set_defaults(func=self.list_configs)
+
+        reset = subs.add_parser(
+            'reset', help='Reset a configuration option to its default value')
+        reset.add_argument('namespace', help='The configuration namespace')
+        reset.add_argument('key', help='The configuration key')
+        reset.set_defaults(func=self.reset_config)
+
+        set = subs.add_parser(
+            'set', help='Set the value of a configuration option')
+        set.add_argument('namespace', help='The configuration namespace')
+        set.add_argument('key', help='The configuration key')
+        set.add_argument('value', help='The new value')
+        set.set_defaults(func=self.set_config)
+
+        self.parser = parser
+
+    def handle(self, *_, **options):
+        if 'func' not in options:
+            self.parser.print_help()
+            self.parser.exit(1)
+
+        options['func'](options)
+
+    def _evaluate_value(self, value):
+        if value in ('true', 'false'):
+            # Special case booleans for a better UX
```

Comment:
One thing to keep in mind is that this is a CLI, and as such all it can receive as input are strings.

We want to transform (`eval`uate) those strings into Python objects.

So if the admin uses the command `config set foo bar 1` we can't really know whether they meant the integer `1`, the string `'1'`, or the boolean `True`.

The way to deal with this ambiguity IMHO is to avoid it at all cost. That means assuming that `1` will always be an integer, period.

That's why to set a config to the string `'1'` you'd need to use the command as follows:

```
$ python manage.py config set server site-name "'1'"
```

The first line of quotes get completely swallowed by the shell (at least that's the case with bash) so if you used `'1'` or `"1"` the `config.py` code only gets `'1'` (i.e a one-character long string containing only the character `1`) which gets evaluated as the integer `1`. Passing the parameter as above means that `config.py` gets `"'1'"` (i.e a thee-character long string containing a single quote, the character `1` then another single quote), which gets evaluated as a string properly.

If we wanted to let the admin type `python manage.py config set server site-name 1` and magically evaluate that `1` as a string, we'd need to peek into the configuration registry to find the expected type, and then try to evaluate it into something of that type.

Sure, that's easy for the parameter `1`: we can evaluate it as a string simply with `str(value)`. Now if the expected type is a list and `config.py` receives the string `foo`, should we pass it through `list()` and thus store `['f', 'o', 'o']`? Of course we shouldn't, that would be a case of an admin trying to store an invalid value, and we should error out.

Similarly, if the expected type is a string and the admin called passed the value as `"['en', 'fr', 'zh-hant']"` (the syntax for passing a list), we also shouldn't pass it to `str()` and store that awful string in the database.

So we'd have special cases only for things that evaluate as integer, if the expected type is a string.

Or, as you suggested, if the passed value evaluates as an integer and the expected type is a boolean.

And then we quickly get into special-corner-cases-territory, and soon we'll start wondering where did we go wrong.

On top of that, I really don't think the CLI layer should go and peek inside the registry on its own, except for things like `list` and `help` which are purely about presenting information to the user (i.e read-only).

So no, I don't think we should handle 0/1 as booleans. (I know, I took a slight detour :laughing:  )

As for on/off and yes/no, I don't think they are needed, but I'll accept I might be wrong. That's also why I'd really like to have people who don't read the code to try it out, because it is something we would figure out instantly if they tried setting a boolean value with something else than true/false. I'm not going to mention them in this conversation because reading the thought process I detailed above (and in fact your very question) would influence how they try using it, and I'm really interested in their feedback as first-time users. :smiley: 

This begs the question as to why I added this special case, then. Well, without it the CLI would only accept the strings `True` and `False` as valid for booleans, and I felt that refusing "true" because it doesn't have the proper case would be incredibly sadistic.